# snpArcher-bam-input

<img src="./docs/img/logo.png" alt="snpArcher logo" height="300"/>


snpArcher is a reproducible workflow optimized for nonmodel organisms and comparisons across datasets, built on the [Snakemake](https://snakemake.readthedocs.io/en/stable/index.html#) workflow management system. It provides a streamlined approach to dataset acquisition, variant calling, quality control, and downstream analysis.

This **modified** version of snpArcher is designed to accept **BAM files as input**, rather than raw sequencing reads. It assumes that the BAM files have been pre-processed (e.g., aligned to a reference genome, sorted, and indexed with **duplicates marked**) and are ready for variant calling. Since this is an unofficial fork of snpArcher, please refer to the original [snpArcher repository](https://github.com/harvardinformatics/snpArcher/tree/main) for the main documentation, installation instructions, and usage guidelines.

### Usage
For usage instructions and complete documentation, please visit our [docs](https://snparcher.readthedocs.io/en/latest/).

In addition to the standard usage instructions, the sample list input file must be modified to include the paths to the BAM files. An example of the sample list format is provided below (with an example in `config/example_samples.csv`):

```
BioSample,Run,LibraryName,refGenome,refPath,bam
Sample_1,external_bam,lib1,ref_genome,/path/to/ref/genome.fa,/path/to/bams/sample_1.bam
Sample_2,external_bam,lib1,ref_genome,/path/to/ref/genome.fa,/path/to/bams/sample_2.bam
Sample_3,external_bam,lib1,ref_genome,/path/to/ref/genome.fa,/path/to/bams/sample_3.bam
```

In addition, the `mark_duplicates:` flag in your `config.yaml` file must be set to `False`, as the input alignment files are expected to have duplicates already marked.

### Datasets generated by snpArcher
A number of resequencing datasets have been run with snpArcher generating consistent variant calls, available via [Globus](https://www.globus.org/) in the [Comparative Population Genomics Data collection](https://app.globus.org/file-manager?origin_id=a6580c44-09fd-11ee-be16-195c41bc0be4&origin_path=%2F). Details of data processing are described [in our manuscript](https://www.biorxiv.org/content/10.1101/2023.06.22.546168v1). If you use any of these datasets in your projects, please cite both the [snpArcher paper](https://www.biorxiv.org/content/10.1101/2023.06.22.546168v1) and the original data producers.

### Citing snpArcher
- Cade D Mirchandani, Allison J Shultz, Gregg W C Thomas, Sara J Smith, Mara Baylis, Brian Arnold, Russ Corbett-Detig, Erik Enbody, Timothy B Sackton, A fast, reproducible, high-throughput variant calling workflow for population genomics, Molecular Biology and Evolution, 2023;, msad270, https://doi.org/10.1093/molbev/msad270
- Also, make sure to cite the tools you used within snpArcher.
